### PR TITLE
Add command to join solo 3v3 rated

### PR DESCRIPTION
**Pull request by laasker**
## Changes Proposed:
- 
- 

## Issues Addressed:

- Closes

## Tests Performed:

-  Tested ingame, works good, didn't found any issue
- 

## How to Test the Changes:

1.
2.
3.

### CODE REVIEW

File: src/solo3v3_sc.cpp [hunk 1]
```diff
@@ -333,7 +336,7 @@ bool NpcSolo3v3::JoinQueueArena(Player* player, Creature* creature, bool isRated
 
 bool NpcSolo3v3::CreateArenateam(Player* player, Creature* creature)
 {
-    if (!player || !creature)
+    if (!player/* || !creature*/)
```
Comment (Helias) [inline src/solo3v3_sc.cpp:339]: you can delete the !creature

File: src/solo3v3_sc.cpp [hunk 2]
```diff
@@ -259,8 +259,11 @@ bool NpcSolo3v3::JoinQueueArena(Player* player, Creature* creature, bool isRated
     uint32 arenaRating = 0;
     uint32 matchmakerRating = 0;
 
-    // ignore if we already in BG or BG queue
-    if (player->InBattleground())
+    // ignore if we already in BG, Arena or Arena queue
+    if (player->InBattleground() || player->InArena() || player->InBattlegroundQueueForBattlegroundQueueType((BattlegroundQueueTypeId)BATTLEGROUND_QUEUE_2v2) ||
+        player->InBattlegroundQueueForBattlegroundQueueType((BattlegroundQueueTypeId)BATTLEGROUND_QUEUE_3v3) ||
+        player->InBattlegroundQueueForBattlegroundQueueType((BattlegroundQueueTypeId)BATTLEGROUND_QUEUE_5v5) ||
+        player->InBattlegroundQueueForBattlegroundQueueType((BattlegroundQueueTypeId)BATTLEGROUND_QUEUE_3v3_SOLO))
```
Comment (Helias) [inline src/solo3v3_sc.cpp:266]: ideally we should add the 1v1 too, but maybe under conf

File: conf/arena_3v3_solo_queue.conf.dist [hunk 3]
```diff
@@ -9,6 +9,8 @@
 #
 
 Solo.3v3.Enable = 1
+Solo.3v3.EnableCommand = 1
+Solo.3v3.EnableTestingCommand = 1
```
Comment (Helias) [inline conf/arena_3v3_solo_queue.conf.dist:13]: I would disable the test command by default

File: src/solo3v3_sc.cpp [hunk 4]
```diff
@@ -249,7 +249,7 @@ bool NpcSolo3v3::ArenaCheckFullEquipAndTalents(Player* player)
 
 bool NpcSolo3v3::JoinQueueArena(Player* player, Creature* creature, bool isRated)
 {
-    if (!player || !creature)
+    if (!player/* || !creature*/)
```
Comment (Helias) [inline src/solo3v3_sc.cpp:252]: you can delete the !creature

File: src/solo3v3_sc.h [hunk 5]
```diff
@@ -145,6 +144,154 @@ class Arena_SC : public ArenaScript
...
+                    if (SoloCommand.JoinQueueArena(currentPlayer, nullptr, true))
+                    {
+                        handler->PSendSysMessage("Player {} has joined the solo 3v3 arena queue.", currentPlayer->GetName().c_str());
+                    }
+                    else
+                    {
+                        handler->PSendSysMessage("Failed to join queue for player {}.", currentPlayer->GetName().c_str());
+                    }
+                }
+            }
```
Comment (Helias) [inline src/solo3v3_sc.h:287]:
```suggestion
        if (!sConfigMgr->GetOption<bool>("Solo.3v3.EnableTestingCommand", true))
        {
            ChatHandler(player->GetSession()).SendSysMessage("Solo 3v3 Arena testing command is disabled.");
            return false;
        }

        if (!sConfigMgr->GetOption<bool>("Solo.3v3.Enable", true))
        {
            ChatHandler(player->GetSession()).SendSysMessage("Solo 3v3 Arena is disabled.");
            return false;
        }

        NpcSolo3v3 SoloCommand;
        for (auto& pair : ObjectAccessor::GetPlayers())
        {
            Player* currentPlayer = pair.second;
            if (currentPlayer)
            {
                if (currentPlayer->IsInCombat())
                {
                    handler->PSendSysMessage("Player {} can't be in combat.", currentPlayer->GetName().c_str());
                    continue;
                }

                if (currentPlayer->HasAura(26013) && (sConfigMgr->GetOption<bool>("Solo.3v3.CastDeserterOnAfk", true) || sConfigMgr->GetOption<bool>("Solo.3v3.CastDeserterOnLeave", true)))
                {
                    WorldPacket data;
                    sBattlegroundMgr->BuildGroupJoinedBattlegroundPacket(&data, ERR_GROUP_JOIN_BATTLEGROUND_DESERTERS);
                    currentPlayer->GetSession()->SendPacket(&data);
                    continue;
                }

                uint32 minLevel = sConfigMgr->GetOption<uint32>("Solo.3v3.MinLevel", 80);
                if (currentPlayer->GetLevel() < minLevel)
                {
                    handler->PSendSysMessage("Player {} needs level {}+ to join solo arena.", player->GetName().c_str(), minLevel);
                    continue;
                }

                if (!currentPlayer->GetArenaTeamId(ARENA_SLOT_SOLO_3v3)) // ARENA_SLOT_SOLO_3v3 | ARENA_TEAM_SOLO_3v3
                {
                    if (!SoloCommand.CreateArenateam(currentPlayer, nullptr))
                        continue;
                }
                else
                {
                    if (!SoloCommand.ArenaCheckFullEquipAndTalents(currentPlayer))
                        continue;

                    if (SoloCommand.JoinQueueArena(currentPlayer, nullptr, true))
                        handler->PSendSysMessage("Player {} has joined the solo 3v3 arena queue.", currentPlayer->GetName().c_str());
                    else
                        handler->PSendSysMessage("Failed to join queue for player {}.", currentPlayer->GetName().c_str());
                }
            }
```

File: src/solo3v3_sc.h [hunk 6]
```diff
@@ -145,6 +144,152 @@ class Arena_SC : public ArenaScript
...
+            if (!SoloCommand.ArenaCheckFullEquipAndTalents(player))
+            {
+                return false;
+            }
+            if (SoloCommand.JoinQueueArena(player, nullptr, true))
+            {
+                handler->PSendSysMessage("You have joined the solo 3v3 arena queue.");
+            }
+        }
+
```
Comment (Helias) [inline src/solo3v3_sc.h:219]:
```suggestion
        if (!sConfigMgr->GetOption<bool>("Solo.3v3.EnableCommand", true))
        {
            ChatHandler(player->GetSession()).SendSysMessage("Solo 3v3 Arena command is disabled.");
            return false;
        }

        if (!sConfigMgr->GetOption<bool>("Solo.3v3.Enable", true))
        {
            ChatHandler(player->GetSession()).SendSysMessage("Solo 3v3 Arena is disabled.");
            return false;
        }

        if (player->IsInCombat())
        {
            ChatHandler(player->GetSession()).SendSysMessage("Can't be in combat.");
            return false;
        }

        NpcSolo3v3 SoloCommand;
        if (player->HasAura(26013) && (sConfigMgr->GetOption<bool>("Solo.3v3.CastDeserterOnAfk", true) || sConfigMgr->GetOption<bool>("Solo.3v3.CastDeserterOnLeave", true)))
        {
            WorldPacket data;
            sBattlegroundMgr->BuildGroupJoinedBattlegroundPacket(&data, ERR_GROUP_JOIN_BATTLEGROUND_DESERTERS);
            player->GetSession()->SendPacket(&data);
            return false;
        }

        uint32 minLevel = sConfigMgr->GetOption<uint32>("Solo.3v3.MinLevel", 80);
        if (player->GetLevel() < minLevel)
        {
            ChatHandler(player->GetSession()).PSendSysMessage("You need level {}+ to join solo arena.", minLevel);
            return false;
        }

        if (!player->GetArenaTeamId(ARENA_SLOT_SOLO_3v3))
        {
            // create solo3v3 team if player don't have it
            if (!SoloCommand.CreateArenateam(player, nullptr))
                return false;
        }
        else
        {
            if (!SoloCommand.ArenaCheckFullEquipAndTalents(player))
                return false;

            if (SoloCommand.JoinQueueArena(player, nullptr, true))
                handler->PSendSysMessage("You have joined the solo 3v3 arena queue.");

        }

```

File: data/sql/db-world/as_2024_03_17_00.sql [hunk 7]
```diff
@@ -21,4 +21,8 @@ INSERT INTO `npc_text` (`ID`, `text0_0`) VALUES (@entryNpcText1, @text1);
 SET @entryNpcText2 := 1000004;
 SET @text2 := '                                |TInterface/ICONS/achievement_bg_killXEnemies_generalsroom:25|t$B                        Solo 3v3 Arena$B$B              Melee Caster Healer [|cffff0000off|r]';
 DELETE FROM `npc_text` WHERE `ID` = @entryNpcText;
-INSERT INTO `npc_text` (`ID`, `text0_0`) VALUES (@entryNpcText2, @text2);
\ No newline at end of file
+INSERT INTO `npc_text` (`ID`, `text0_0`) VALUES (@entryNpcText2, @text2);
+
+-- Command
+INSERT INTO `command` (`name`, `security`, `help`) VALUES ('qsolo', 0, '');
+INSERT INTO `command` (`name`, `security`, `help`) VALUES ('testqsolo', 4, '');
```
Comment (Helias) [inline data/sql/db-world/as_2024_03_17_00.sql:28]: when you put an INSERT, put always a DELETE before https://www.azerothcore.org/wiki/sql-standards

Comment (Helias) [reply data/sql/db-world/as_2024_03_17_00.sql:28]:
```sql
DELETE FROM `command` WHRE `name` IN ('qsolo', 'testqsolo');
INSERT INTO `command` (`name`, `security`, `help`) VALUES
('qsolo', 0, ''),
('testqsolo', 4, '');
```

File: src/solo3v3_sc.cpp (hunk 2 above)
Comment (laasker) [reply src/solo3v3_sc.cpp:266]: true, i forgot about that one

File: data/sql/db-world/as_2024_03_17_00.sql [hunk 8]
```diff
@@ -21,4 +21,10 @@ INSERT INTO `npc_text` (`ID`, `text0_0`) VALUES (@entryNpcText1, @text1);
 SET @entryNpcText2 := 1000004;
 SET @text2 := '                                |TInterface/ICONS/achievement_bg_killXEnemies_generalsroom:25|t$B                        Solo 3v3 Arena$B$B              Melee Caster Healer [|cffff0000off|r]';
 DELETE FROM `npc_text` WHERE `ID` = @entryNpcText;
-INSERT INTO `npc_text` (`ID`, `text0_0`) VALUES (@entryNpcText2, @text2);
\ No newline at end of file
+INSERT INTO `npc_text` (`ID`, `text0_0`) VALUES (@entryNpcText2, @text2);
+
+-- Command
+DELETE FROM `command` WHRE `name` IN ('qsolo', 'testqsolo');
```
Comment (Helias) [inline data/sql/db-world/as_2024_03_17_00.sql:27]: sorry, my fault
```suggestion
DELETE FROM `command` WHERE `name` IN ('qsolo', 'testqsolo');
```

File: data/sql/db-world/as_2024_03_17_00.sql [hunk 9]
```diff
@@ -21,4 +21,10 @@ INSERT INTO `npc_text` (`ID`, `text0_0`) VALUES (@entryNpcText1, @text1);
...
 DELETE FROM `npc_text` WHERE `ID` = @entryNpcText;
-INSERT INTO `npc_text` (`ID`, `text0_0`) VALUES (@entryNpcText2, @text2);
\ No newline at end of file
+INSERT INTO `npc_text` (`ID`, `text0_0`) VALUES (@entryNpcText2, @text2);
+
+-- Command
+DELETE FROM `command` WHERE `name` IN ('qsolo', 'testqsolo');
+INSERT INTO `command` (`name`, `security`, `help`) VALUES
+('qsolo', 0, ''),
+('testqsolo', 4, '');
```
Comment (Helias) [inline data/sql/db-world/as_2024_03_17_00.sql:30]: adding a newline at the end
```suggestion
('testqsolo', 4, '');

```

**Review by Helias — APPROVED** [review]

File: src/solo3v3_sc.h [hunk 10]
```diff
@@ -145,6 +144,152 @@ class Arena_SC : public ArenaScript
...
+        static ChatCommandTable SoloCommandTable =
+        {
+            { "qsolo",     HandleQueueSoloArena,         SEC_PLAYER,        Console::No },
+            { "testqsolo", HandleQueueSoloArenaTesting,  SEC_ADMINISTRATOR, Console::No }
+        };
+
+        return SoloCommandTable;
+    }
+
+    static bool HandleQueueSoloArena(ChatHandler* handler, const char* args)
```
Comment (Helias) [inline src/solo3v3_sc.h:165]:
```suggestion
    static bool HandleQueueSoloArena(ChatHandler* handler, const char* /*args*/)
```

**Comment by Helias** [comment]
```
/home/runner/work/mod-arena-3v3-solo-queue/mod-arena-3v3-solo-queue/modules/mod-arena-3v3-solo-queue/src/solo3v3_sc.cpp:250:59: fatal error: unused parameter 'creature' [-Wunused-parameter]
  250 | bool NpcSolo3v3::JoinQueueArena(Player* player, Creature* creature, bool isRated)
      |                                                           ^
1 error generated.
```
https://github.com/azerothcore/mod-arena-3v3-solo-queue/actions/runs/10848687893/job/30106330449?pr=17

File: src/solo3v3_sc.cpp [hunk 11]
```diff
@@ -249,7 +249,7 @@ bool NpcSolo3v3::ArenaCheckFullEquipAndTalents(Player* player)
 
 bool NpcSolo3v3::JoinQueueArena(Player* player, Creature* creature, bool isRated)
```
Comment (Helias) [inline src/solo3v3_sc.cpp:250]:
```suggestion
bool NpcSolo3v3::JoinQueueArena(Player* player, Creature* /*creature*/, bool isRated)
```

File: src/solo3v3_sc.cpp [hunk 12]
```diff
@@ -333,7 +337,7 @@ bool NpcSolo3v3::JoinQueueArena(Player* player, Creature* creature, bool isRated
 
 bool NpcSolo3v3::CreateArenateam(Player* player, Creature* creature)
```
Comment (Helias) [inline src/solo3v3_sc.cpp:338]:
```suggestion
bool NpcSolo3v3::CreateArenateam(Player* player, Creature* /*creature*/)
```

**Pull request merged**